### PR TITLE
start search fix

### PR DESCRIPTION
**Pull request by sarracini**
First pass at fixing the search bar

### CODE REVIEW

**Comment by sarracini** [comment]
contentSearchUI.js desperately needs to be jshint'ed

**Comment by sarracini** [comment]
@marcoscaceres @oyiptong @emtwo r?

File: css/contentSearchUI.css [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+.contentSearchSuggestionRow > td > .historyIcon {
+  margin-right: 8px;
+  margin-bottom: -3px;
+}
+
+.contentSearchSuggestionTable .historyIcon {
+  width: 16px;
+  height: 16px;
+  display: inline-block;
+  background-image: url("chrome://browser/skin/search-history-icon.svg#search-history-icon");
```
Comment (marcoscaceres) [inline css/contentSearchUI.css]: needs to be remote... same for the others.

File: js/contentSearchUI.js [hunk 2]
```diff
@@ -0,0 +1,898 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*global gNewTab, KeyboardEvent*/
+
+"use strict";
+(function(exports) {
+
+  const MAX_DISPLAYED_SUGGESTIONS = 6;
+  const SUGGESTION_ID_PREFIX = "searchSuggestion";
+  const ONE_OFF_ID_PREFIX = "oneOff";
```
Comment (marcoscaceres) [inline js/contentSearchUI.js]: this is only used once, so just inline it.

File: js/contentSearchUI.js [hunk 3]
```diff
@@ -0,0 +1,898 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*global gNewTab, KeyboardEvent*/
+
+"use strict";
+(function(exports) {
+
+  const MAX_DISPLAYED_SUGGESTIONS = 6;
+  const SUGGESTION_ID_PREFIX = "searchSuggestion";
```
Comment (marcoscaceres) [inline js/contentSearchUI.js]: this probably doesn't need a const

File: js/contentSearchUI.js [hunk 4]
```diff
@@ -0,0 +1,898 @@
...
+  const ONE_OFF_ID_PREFIX = "oneOff";
+
+  const HTML_NS = "http://www.w3.org/1999/xhtml";
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:9]: delete this, and use `document.createElement()` instead.

File: js/contentSearchUI.js [hunk 5]
```diff
@@ -0,0 +1,898 @@
...
+   *        top left of the page.
+   * @param {healthReportKey} healthReportKey
+   *        This will be sent with the search data for FHR to record the search.
+   * @param {String} searchPurpose
+   *        Sent with search data, see nsISearchEngine.getSubmission.
+   * @param {String} idPrefix
+   *        The IDs of elements created by the object will be prefixed with this
+   *        string.
+   */
+  function ContentSearchUIController(inputElement, tableParent, healthReportKey,
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:36]: Why does this need a constructor?

File: js/contentSearchUI.js [hunk 6]
```diff
@@ -0,0 +1,898 @@
...
+      return this._defaultEngine;
+    },
+
+    set defaultEngine(engine) {
+      let icon;
+      if (engine.iconBuffer) {
+        icon = this._getFaviconURIFromBuffer(engine.iconBuffer);
+      } else {
+        icon = this._getImageURIForCurrentResolution(
+          "chrome://mozapps/skin/places/defaultFavicon.png");
```
Comment (marcoscaceres) [inline js/contentSearchUI.js]: this need to point to be `images/defaultFavicon.png`

Comment (marcoscaceres) [inline js/contentSearchUI.js]: We probably need to add this file to the remote site.

File: js/contentSearchUI.js [hunk 7]
```diff
@@ -0,0 +1,898 @@
...
+    },
+
+    // The selectedIndex is the index of the element with the "selected" class in
+    // the list obtained by concatenating the suggestion rows, one-off buttons, and
+    // search settings button.
+    get selectedIndex() {
+      let allElts = [...this._suggestionsList.children,
+        ...this._oneOffButtons,
+        document.getElementById("contentSearchSettingsButton")
+      ];
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:117]: heh, fancy... Maybe:

``` JS
let allElts = this._suggestionsList.children
  .concat(this._oneOffButtons)
  .concat(document.getElementById("contentSearchSettingsButton"));
```

File: js/contentSearchUI.js [hunk 8]
```diff
@@ -0,0 +1,898 @@
...
+      for (let i = 0; i < allElts.length; ++i) {
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:118]: Replace with: 

``` JS
return allEts.findIndex(elt => elt.classList.contains("selected"));
```

File: js/contentSearchUI.js [hunk 9]
```diff
@@ -0,0 +1,898 @@
...
+      }
+      return -1;
+    },
+
+    set selectedIndex(idx) {
+      // Update the table's rows, and the input when there is a selection.
+      this._table.removeAttribute("aria-activedescendant");
+      this.input.removeAttribute("aria-activedescendant");
+
+      let allElts = [...this._suggestionsList.children,
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:132]: As above. Also, repeated code... maybe need a function.

File: js/contentSearchUI.js [hunk 10]
```diff
@@ -0,0 +1,898 @@
...
+        ...this._oneOffButtons,
+        document.getElementById("contentSearchSettingsButton")
+      ];
+      // If we are selecting a suggestion and a one-off is selected, don't deselect it.
+      let excludeIndex = idx < this.numSuggestions && this.selectedButtonIndex > -1 ?
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:137]: please split this up

File: js/contentSearchUI.js [hunk 11]
```diff
@@ -0,0 +1,898 @@
...
+        this.numSuggestions + this.selectedButtonIndex : -1;
+      for (let i = 0; i < allElts.length; ++i) {
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:139]: This is ugly :( Can we split this up?

File: js/contentSearchUI.js [hunk 12]
```diff
@@ -0,0 +1,898 @@
...
+    this._getStrings();
+  }
+
+  ContentSearchUIController.prototype = {
+
+    // The timeout (ms) of the remote suggestions.  Corresponds to
+    // SearchSuggestionController.remoteTimeout.  Uses
+    // SearchSuggestionController's default timeout if falsey.
+    remoteTimeout: undefined,
+    _oneOffButtons: [],
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:70]: We can do better here... thinking a live node list or a smarter getter.

File: js/contentSearchUI.js [hunk 13]
```diff
@@ -0,0 +1,898 @@
...
+        }
+      }
+    },
+
+    get selectedButtonIndex() {
+      let elts = [...this._oneOffButtons,
+        document.getElementById("contentSearchSettingsButton")
+      ];
+      for (let i = 0; i < elts.length; ++i) {
+        if (elts[i].classList.contains("selected")) {
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:158]: As above. use `.filter()`

Comment (marcoscaceres) [inline js/contentSearchUI.js:158]: Err... I mean `.indexOf`

File: js/contentSearchUI.js [hunk 14]
```diff
@@ -0,0 +1,898 @@
...
+          return i;
+        }
+      }
+      return -1;
+    },
+
+    set selectedButtonIndex(idx) {
+      let elts = [...this._oneOffButtons,
+        document.getElementById("contentSearchSettingsButton")
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:167]: We should store this somewhere. Not great to be doing it over and over.

File: js/contentSearchUI.js [hunk 15]
```diff
@@ -0,0 +1,898 @@
...
+      ];
+      for (let i = 0; i < elts.length; ++i) {
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:169]: Maybe:

``` JS
var selected = elts[idx]
selected.classList.add("selected");
selected.setAttribute("aria-selected", "true");
elts.filter(elem => elem !== selected)
  .forEach(elt => {
    elt.classList.remove("selected");
    elt.setAttribute("aria-selected", "false");
  });

```

File: js/contentSearchUI.js [hunk 16]
```diff
@@ -0,0 +1,898 @@
...
+    },
+
+    addInputValueToFormHistory: function() {
+      gNewTab.sendToBrowser("NewTab:AddFormHistoryEntry", {
+        entry: this.input.value
+      });
+    },
+
+    handleEvent: function(event) {
+      this["_on" + event.type[0].toUpperCase() + event.type.substr(1)](event);
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:225]: :(

File: js/contentSearchUI.js [hunk 17]
```diff
@@ -0,0 +1,898 @@
...
+        return;
+      case event.DOM_VK_ESCAPE:
+        if (!this._table.hidden) {
+          this._hideSuggestions();
+        }
+        return;
+      default:
+        return;
+      }
+
```
Comment (marcoscaceres) [inline js/contentSearchUI.js:384]: Is this code reachable?

File: js/search.js [hunk 18]
```diff
@@ -0,0 +1,17 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*globals ContentSearchUIController*/
+"use strict";
```
Comment (marcoscaceres) [inline js/search.js:5]: I don't know if we need this file.... we should see if we can export `ContentSearchUIController` without needing a constructor.

**Comment by marcoscaceres** [comment]
We can take this code as is, tho it could be improved (@sarracini, I know you didn't write most of this code).... I guess we can just focus on the page/gecko communication for now, so feel free to ignore my other suggestions.

**Comment by oyiptong** [comment]
Also, I'd like to see some tests for `contentSearchUI.js`.
It is an incredibly valuable piece of functionality and we'll want to track any regression.

**Comment by marcoscaceres** [comment]
@sarracini need rebase. Let me know if you want me to do rebase.

**Comment by sarracini** [comment]
@marcoscaceres rebased and squashed the commits :)

**Comment by marcoscaceres** [comment]
@sarracini, thanks!

**Comment by marcoscaceres** [comment]
@oyiptong, I think we should merge this and make it a separate bug to migrate the tests over. It will be significant work to migrate the tests over, and I don't think we should block on it. Ursula's additions only send messages back and forth.

**Comment by oyiptong** [comment]
i'm afraid we will forget. but i'll settle with some sanity tests

i.e. no need for deep corner-case tests, but at least that the main functionality works and that the main failure scenarios are handled

**Comment by oyiptong** [comment]
also: no need for DOM manipulation tests if we can cut somewhere

File: js/contentSearchUI.js [hunk 19]
```diff
@@ -0,0 +1,894 @@
...
+          this.input.value + "</span>";
+        html = this._strings.searchForKeywordsWith.replace("%S", html);
+        searchWithHeader.innerHTML = html;
+        return;
+      }
+      searchWithHeader.appendChild(document.createTextNode(this._strings.searchWithHeader));
+    },
+
+    _speculativeConnect: function() {
+      if (this.defaultEngine) {
```
Comment (marcoscaceres) [inline js/contentSearchUI.js]: Maybe:

``` JS
let preConnector = document.getElementById("preconnector");
if (this.defaultEngine && this.defaultEngine.preconnectOrigin !== preConnector.href)  {
   preConnector.href = this.defaultEngine.preconnectOrigin;
}
```

File: index.html [hunk 20]
```diff
@@ -8,6 +8,8 @@
     <meta charset="utf-8">
     <title>New Tab</title>
     <link rel="stylesheet" href="css/newTab.css">
+    <link rel="stylesheet" href="css/contentSearchUI.css">
+    <link rel="preconnect" id="preconnector" crossorigin="anonymous">
```
Comment (marcoscaceres) [inline index.html]: I wonder if an engine should manage its own preconnector. That is, when the default engine is set, we create this element and add it to the DOM. Then, if the default engine is clobbered, it removes this from the DOM.

Comment (sarracini) [inline index.html]: happy to create this element on the fly, as speculative connect gets called, but not sure if there's ever a time when the default engine is clobbered. we could remove it from the DOM after the search is done though, since the speculative connect isn't needed after the search is performed.
